server/src/main/java/org/elasticsearch/ReleaseVersions.java

```diff
@@ -0,0 +1,128 @@
+import java.util.Collections;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.ServiceLoader;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Pattern;
+
+public class ReleaseVersions {
+
+    private static final Logger Log = LogManager.getLogger(ReleaseVersions.class);
+
+    private static final boolean USES_VERSIONS;
+
+    static {
+        USES_VERSIONS = ExtensionLoader.loadSingleton(ServiceLoader.load(BuildExtension.class))
+            .map(BuildExtension::hasReleaseVersioning)
+            .orElse(true);
+    }
+
+    private static final Map<Class<?>, NavigableMap<Integer, String>> RESOLVED_VERSIONS = new ConcurrentHashMap<>();
+
+    private static final Pattern COMMENT_LINE = Pattern.compile("^\\h*#");
+    private static final Pattern VERSION_LINE = Pattern.compile("(\\d+\\.\\d+\\.\\d+),(\\d+)(\\h*#.*)?");
+
+    static NavigableMap<Integer, String> readVersionsFile(Class<?> versionContainer) {
+        String versionsFileName = versionContainer.getSimpleName() + ".csv";
+        URL versionsFile = versionContainer.getResource(versionsFileName);
+        if (versionsFile == null) {
+            Log.error("Could not find versions file for class [{}]", versionContainer);
+            return Collections.emptyNavigableMap();
+        }
+
+        try (
+            BufferedReader reader = new BufferedReader(
+                new InputStreamReader(versionContainer.getResourceAsStream(versionsFileName), StandardCharsets.UTF_8)
+            )
+        ) {
+            NavigableMap<Integer, String> versions = new TreeMap<>();
+            String line;
+            while ((line = reader.readLine()) != null) {
+                if (COMMENT_LINE.matcher(line).find()) continue; // a comment
+
+                var matcher = VERSION_LINE.matcher(line);
+                if (matcher.matches() == false) {
+                    Log.warn("Incorrect format for line [{}] in [{}]", line, versionsFile);
+                    continue;
+                }
+                try {
+                    Integer id = Integer.valueOf(matcher.group(2));
+                    String version = matcher.group(1);
+                    String existing = versions.putIfAbsent(id, version);
+                    if (existing != null) {
+                        Log.warn("Duplicate id [{}] for versions [{}, {}] in [{}]", id, existing, version, versionsFile);
+                    }
+                } catch (NumberFormatException e) {
+                    // cannot happen??? regex is wrong...
+                    assert false : "Regex allowed non-integer id through: " + e;
+                    Log.warn("Incorrect format for line [{}] in [{}]", line, versionsFile, e);
+                }
+            }
+
+            return Collections.unmodifiableNavigableMap(versions);
+        } catch (Exception e) {
+            Log.error("Could not read versions file [{}]", versionsFile, e);
```

Comment:
At the moment, this is only first scanned when it is requested, so this would cause failures at random points - hence why it fails in non-fatal ways.

We could have it eagerly load all the version mappings on startup, then it could fail outright if the files are incorrect. That would give more predictability

Reply:
I've change it to do this - eagerly load the files on startup, and so blowing up if the file is malformed